### ISSUE TITLE
Indexing into torch.nn.Modulelist with indices >= 2 returns "torch._dynamo.exc.InternalTorchDynamoError: SymNodeVariable() is not a constant"

### ISSUE THREAD

**Issue by Adam27X** [pytorch/pytorch#114171]
### 🐛 Describe the bug

I have a series of torch.nn.Sequential objects in a model that loops over them. My model happens to have four of these objects and it has a function that takes in an index to choose which of the Sequential objects to use for that particular invocation of the function. If I compile this function and pass 0 or 1 I can generate an Fx Graph just fine, but if I pass 2 or 3 I end up with the stack trace below. This issue seems related to unsound 0/1 specialization with the quirk that it happens to work for 0/1 and not other values.

This issue may also be related to: https://github.com/pytorch/pytorch/issues/103619

### Error logs

Traceback (most recent call last):
  File "sym_node_variable_not_a_constant.py", line 45, in <module>
    y2 = forward_aot(x,2) #Fails
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 328, in _fn
    return fn(*args, **kwargs)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 490, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 133, in _fn
    return fn(*args, **kwargs)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 389, in _convert_frame_assert
    return _compile(
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 586, in _compile
    raise InternalTorchDynamoError(str(e)).with_traceback(
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 569, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 189, in time_wrapper
    r = func(*args, **kwargs)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 491, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 458, in transform
    tracer.run()
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2069, in run
    super().run()
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 719, in run
    and self.step()
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 683, in step
    getattr(self, inst.opname)(inst)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 392, in wrapper
    return inner_fn(self, inst)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 168, in impl
    self.push(fn_var.call_function(self, self.popn(nargs), {}))
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 618, in call_function
    result = handler(tx, *args, **kwargs)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 950, in call_getitem
    return args[0].call_method(tx, "__getitem__", args[1:], kwargs)
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/variables/nn_module.py", line 590, in call_method
    key = args[0].as_python_constant()
  File "/gdn/centos7/0001/x3/prefixes/pytorch/2.1.1-cuda12.1.1-02c7__17ecf4b7bca0/lib/python3.10/site-packages/torch/_dynamo/variables/base.py", line 225, in as_python_constant
    raise NotImplementedError(f"{self} is not a constant")
torch._dynamo.exc.InternalTorchDynamoError: SymNodeVariable() is not a constant

from user code:
   File "sym_node_variable_not_a_constant.py", line 29, in interact
    return self.interaction_networks[cycle](x)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information

You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

### Minified repro

    import torch

    from typing import List
    from torch._dynamo.backends.common import aot_autograd
    from functorch.compile import make_boxed_func, aot_function

    device = torch.device("cuda")

    def fcnn(in_dim, out_dim, hidden_dim, activation=torch.nn.GELU):
        layers = [torch.nn.Linear(in_dim, hidden_dim, device=device), activation(), torch.nn.Linear(hidden_dim, out_dim, device=device)]
        return torch.nn.Sequential(*layers)

    class testmodel(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.interaction_networks = torch.nn.ModuleList([fcnn(262,1174,400) for _ in range(4)])

        def interact(self, x, cycle):
            return self.interaction_networks[cycle](x)

    model = testmodel()

    graph_modules = []
    def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
        graph_modules.append(gm)
        return make_boxed_func(gm.forward)

    my_compiler = aot_autograd(fw_compiler=my_compiler,bw_compiler=my_compiler)

    forward_aot = torch.compile(model.interact, fullgraph=True, dynamic=True, backend=my_compiler)

    x = torch.rand([111,262], device=device)
    #y0 = forward_aot(x,0) #Passes
    #y1 = forward_aot(x,1) #Passes
    y2 = forward_aot(x,2) #Fails
    #y3 = forward_aot(x,3) #Fails

    for graphmodule in graph_modules:
        graphmodule.print_readable()

### Versions

v2.1.1 tag

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by ezyang** [comment]
The simple workaround is to not use dynamic=True

**Comment by Adam27X** [comment]
> The simple workaround is to not use dynamic=True

This workaround unfortunately doesn't work for my particular use case; I _want_ symbolic shapes in the compiled FxGraph. However, I was able to rewrite my compiled function such that it receives `self.interaction_networks[cycle]` as a parameter rather than `cycle` and that allows me to make progress. As always, I appreciate your reply!

**Comment by ezyang** [comment]
Linked PR should fix this too so the code rewrite is not necessary

**Comment by Adam27X** [comment]
> Linked PR should fix this too so the code rewrite is not necessary

Looks like it failed to merge for some reason? I'm not blocked by this so this is low priority, but I am curious about what's going on here.
